Modules/socketmodule.c

```diff
@@ -3069,37 +3088,127 @@ sock_settimeout(PySocketSockObject *s, PyObject *arg)
+        return NULL;
+    }
+
+    if (internal_settimeout(s, timeout) < 0) {
         return NULL;
+    }
+
+    Py_RETURN_NONE;
+}
+
+static int
+internal_settimeout(PySocketSockObject *s, _PyTime_t timeout) {
+    int block;
 
     s->sock_timeout = timeout;
 
-    int block = timeout < 0;
-    /* Blocking mode for a Python socket object means that operations
-       like :meth:`recv` or :meth:`sendall` will block the execution of
-       the current thread until they are complete or aborted with a
-       `TimeoutError` or `socket.error` errors.  When timeout is `None`,
-       the underlying FD is in a blocking mode.  When timeout is a positive
-       number, the FD is in a non-blocking mode, and socket ops are
-       implemented with a `select()` call.
-
-       When timeout is 0.0, the FD is in a non-blocking mode.
-
-       This table summarizes all states in which the socket object and
-       its underlying FD can be:
-
-       ==================== ===================== ==============
-        `gettimeout()`       `getblocking()`       FD
-       ==================== ===================== ==============
-        ``None``             ``True``              blocking
-        ``0.0``              ``False``             non-blocking
-        ``> 0``              ``True``              non-blocking
-    */
+    if (s->sock_family == AF_INET || s->sock_family == AF_INET6) {
+        if (timeout == 0) {
+            block = 0;
+        } else if (timeout < 0) {
+            block = 1;
+
+            #ifdef MS_WINDOWS
+            DWORD zero = 0;
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_SNDTIMEO, (char *)&zero,
+                           sizeof(DWORD)) != 0)
+            #else
+            struct timeval zero = {
+                .tv_sec = 0,
+                .tv_usec = 0,
+            };
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_SNDTIMEO, (char *)&zero,
+                           sizeof(struct timeval)) != 0)
+            #endif
+            {
+                // EINVAL means emote closed the socket fd or shutdown has been
+                // called.
+                if (!CHECK_ERRNO(EINVAL)) {
+                    set_error();
+                    return -1;
+                }
+            }
+            #ifdef MS_WINDOWS
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_RCVTIMEO, (char *)&zero,
+                           sizeof(int)) != 0)
+            #else
+            if (setsockopt(s->sock_fd, SOL_SOCKET, SO_RCVTIMEO, (char *)&zero,
+                           sizeof(struct timeval)) != 0)
+            #endif
+            {
+                // EINVAL means emote closed the socket fd or shutdown has been
+                // called.
+                if (!CHECK_ERRNO(EINVAL)) {
+                    set_error();
+                    return -1;
+                }
+            }
+        } else {
+            block = 1;
+
+            #ifdef MS_WINDOWS
+            DWORD timeout_ms = timeout / 1000000;
```

Comment:
```suggestion
            _PyTime_t timeout_as_ms = _PyTime_AsMilliseconds(timeout, _PyTime_ROUND_FLOOR);
            DWORD timeout_ms = INFINITE;
            if ((DWORD)timeout_as_ms == timeout_as_ms) {
                timeout_ms = (DWORD)timeout_as_ms;
            }
```

Untested, and may be overkill (if we're already certain that `timeout` in ms fits into `DWORD`, then just add an explicit cast), but we can't have the warning.

Reply:
[DWORD](https://learn.microsoft.com/en-us/cpp/assembler/masm/dword?view=msvc-170) is always `uint32_t` and `_PyTime_t` is always `int64_t`. And:

```powershell
PS C:\Users\xxxxx> [int64]::MaxValue / 1000000 -gt [uint32]::MaxValue
True
```

So the convert is must have.